### ISSUE TITLE
is "num_group" implement mobilenet?

### ISSUE THREAD

**Issue by qingzhouzhen** [apache/incubator-mxnet#7064]
Is argumnet "num_group" in mx.sym.Convolution has implemented the article MobileNets: [Efficient Convolutional Neural Networks for Mobile Vision Applications:](https://arxiv.org/abs/1704.04861)?
I have a demo used "num_group" in my repository: [https://github.com/qingzhouzhen/mobilenet-mxnet/blob/master/mobilenet.py](https://github.com/qingzhouzhen/mobilenet-mxnet), I added this symbol to emamples and train it use ImageNet, train speed  is about 56 samples/second, like this:
INFO:root:Epoch[0] Batch [20]   Speed: 56.48 samples/sec        accuracy=0.000744
INFO:root:Epoch[0] Batch [40]   Speed: 56.53 samples/sec        accuracy=0.000781
INFO:root:Epoch[0] Batch [60]   Speed: 56.53 samples/sec        accuracy=0.000000
INFO:root:Epoch[0] Batch [80]   Speed: 56.55 samples/sec        accuracy=0.001563
INFO:root:Epoch[0] Batch [100]  Speed: 56.55 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [120]  Speed: 56.55 samples/sec        accuracy=0.002734
INFO:root:Epoch[0] Batch [140]  Speed: 56.53 samples/sec        accuracy=0.001563
INFO:root:Epoch[0] Batch [160]  Speed: 56.53 samples/sec        accuracy=0.002734
INFO:root:Epoch[0] Batch [180]  Speed: 56.54 samples/sec        accuracy=0.001172
INFO:root:Epoch[0] Batch [200]  Speed: 56.50 samples/sec        accuracy=0.002734
INFO:root:Epoch[0] Batch [220]  Speed: 56.53 samples/sec        accuracy=0.001953
INFO:root:Epoch[0] Batch [240]  Speed: 56.55 samples/sec        accuracy=0.003516
INFO:root:Epoch[0] Batch [260]  Speed: 56.54 samples/sec        accuracy=0.001953
INFO:root:Epoch[0] Batch [280]  Speed: 56.55 samples/sec        accuracy=0.005078
INFO:root:Epoch[0] Batch [300]  Speed: 56.54 samples/sec        accuracy=0.004297
And in another version, I do not use "num_group"(I set it to 1) to do [fully converlution like this,](https://github.com/qingzhouzhen/mobilenet-mxnet/blob/master/mobilenet_normal.py)  train speed is about 93 samples/second, like this:
INFO:root:Epoch[0] Batch [20]   Speed: 92.89 samples/sec        accuracy=0.002604
INFO:root:Epoch[0] Batch [40]   Speed: 93.21 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [60]   Speed: 93.19 samples/sec        accuracy=0.003516
INFO:root:Epoch[0] Batch [80]   Speed: 93.09 samples/sec        accuracy=0.000781
INFO:root:Epoch[0] Batch [100]  Speed: 93.11 samples/sec        accuracy=0.001953
INFO:root:Epoch[0] Batch [120]  Speed: 93.16 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [140]  Speed: 93.16 samples/sec        accuracy=0.001563
INFO:root:Epoch[0] Batch [160]  Speed: 93.20 samples/sec        accuracy=0.003906
INFO:root:Epoch[0] Batch [180]  Speed: 93.35 samples/sec        accuracy=0.003125
INFO:root:Epoch[0] Batch [200]  Speed: 93.74 samples/sec        accuracy=0.001953
INFO:root:Epoch[0] Batch [220]  Speed: 93.65 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [240]  Speed: 93.72 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [260]  Speed: 93.62 samples/sec        accuracy=0.002734
INFO:root:Epoch[0] Batch [280]  Speed: 93.59 samples/sec        accuracy=0.003125
INFO:root:Epoch[0] Batch [300]  Speed: 93.51 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [320]  Speed: 93.52 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [340]  Speed: 93.59 samples/sec        accuracy=0.003906
INFO:root:Epoch[0] Batch [360]  Speed: 93.72 samples/sec        accuracy=0.001953
INFO:root:Epoch[0] Batch [380]  Speed: 93.47 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [400]  Speed: 93.45 samples/sec        accuracy=0.003516
INFO:root:Epoch[0] Batch [420]  Speed: 93.59 samples/sec        accuracy=0.002344
INFO:root:Epoch[0] Batch [440]  Speed: 93.65 samples/sec        accuracy=0.004297
INFO:root:Epoch[0] Batch [460]  Speed: 93.70 samples/sec        accuracy=0.004687
INFO:root:Epoch[0] Batch [480]  Speed: 93.59 samples/sec        accuracy=0.005859
INFO:root:Epoch[0] Batch [500]  Speed: 93.73 samples/sec        accuracy=0.006250
INFO:root:Epoch[0] Batch [520]  Speed: 93.67 samples/sec        accuracy=0.003906
INFO:root:Epoch[0] Batch [540]  Speed: 93.48 samples/sec        accuracy=0.006250
INFO:root:Epoch[0] Batch [560]  Speed: 93.68 samples/sec        accuracy=0.003906
INFO:root:Epoch[0] Batch [580]  Speed: 93.65 samples/sec        accuracy=0.005859
INFO:root:Epoch[0] Batch [600]  Speed: 93.67 samples/sec        accuracy=0.003125
INFO:root:Epoch[0] Batch [620]  Speed: 93.67 samples/sec        accuracy=0.006641
INFO:root:Epoch[0] Batch [640]  Speed: 93.67 samples/sec        accuracy=0.008594
INFO:root:Epoch[0] Batch [660]  Speed: 93.73 samples/sec        accuracy=0.007812
INFO:root:Epoch[0] Batch [680]  Speed: 93.84 samples/sec        accuracy=0.008594
INFO:root:Epoch[0] Batch [700]  Speed: 93.74 samples/sec        accuracy=0.005859
INFO:root:Epoch[0] Batch [720]  Speed: 93.65 samples/sec        accuracy=0.006641

I think I am wrong, mobilenet is to speed up train speed, but my demo have slowed down it?

**Comment by feiyulv** [comment]
Setting the num_group to filter number is equal to the depth-wise convolution. I think your implmentation is right.

**Comment by qingzhouzhen** [comment]
I hava update my problem, @feiyulv

**Comment by feiyulv** [comment]
Have you solved  the accurancy problem?  Slow is the right reuslt, because cudnn doesn't support group convolution, current group conv is done by call cudnn conv `c` times.  'c' is the channel number. Tensorflow has implemented depth-wise conv using cuda kernel, but speed is still not very good.

**Comment by wetliu** [comment]
I tested the speed between TF depth wise convolution with the 1*1 convolution and the normal convolution. The speed of depth wise conv with the 1*1 operator is 2 times as fast as the normal convolution. I think the group trick is similar to the depth-wise convolution, but it seems there is some difference (I am not very sure).

**Comment by xysong1201** [comment]
> Setting the num_group to filter number is equal to the depth-wise convolution. I think your implmentation is right.

Hi, @feiyulv , could you specify how to define the num_group to create depth-convolution? Thank you
